# Lab 3

Katherine Atkins
1001917375

In this lab we will become familiar with distributions, histograms, and functional programming. Do not use numpy or any other library for this lab.

Before that, lets get setup homework submission and submit your previous lab.  

## Working on the Command-line.

It is important for you to learn to work on the command line and to be familiar with the Unix environment (e.g. Linux, Mac OS, or Windows Linux Subsystem). We'll go over working on the command-line in detail later in the course.

You are required to submit your work in this course via GitHub. Today in class, you will setup everything on the command-line.

### Command-line basics

There is plenty of material online that will help you figure out how to do various tasks on the command line. Commands you may need to know today:

* `ls`: lists the contents of the current directory.
* `pwd`: prints the path of the current directory.
* `cd <directory>`: changes your current directory to the specified directory.
* `cd ..`: changes current directory to the previous directory. Basically steps out of the current directory to the directory containing the current directory.
* `mkdir <directory>`: create a new directory with the specified name.
* `rmdir <directory>`: removes the specified directory. Note it has to be empty.
* `rm <filename>`: deletes the specified file.
* `mv <filename 1> <filename 1>`: Moves or renames a file.
* `cp <filename 1> <filename 2>`: copies an file. If you just provide a path to a directory, it copies the file into that directory with the same filename. If you specifiy a new filename, the copy has a new name. For example `cp File.1.txt File.2.txt` creates a copy of `File.1.txt` with the name `File.2.txt`. Meanwhile `cp File.1.txt my_directory`, where `my_directory` is a directory, creates a copy of `File.1.txt` in directory `my_directory` with the name `File.1.txt`.

For reference, here are some example resources I found by googling:

* Paths and Wildcards: https://www.warp.dev/terminus/linux-wildcards
* Basic commands like copy: https://kb.iu.edu/d/afsk
* General introduction to shell: https://github-pages.ucl.ac.uk/RCPSTrainingMaterials/HPCandHTCusingLegion/2_intro_to_shell.html
* Manual pages: https://www.geeksforgeeks.org/linux-man-page-entries-different-types/?ref=ml_lbp
* Chaining commands: https://www.geeksforgeeks.org/chaining-commands-in-linux/?ref=ml_lbp
* Piping: https://www.geeksforgeeks.org/piping-in-unix-or-linux/
* Using sed: https://www.geeksforgeeks.org/sed-command-linux-set-2/?ref=ml_lbp
* Various Unix commands: https://www.geeksforgeeks.org/linux-commands/?ref=lbp
* Cheat sheets:
    * https://www.stationx.net/unix-commands-cheat-sheet/
    * https://cheatography.com/davechild/cheat-sheets/linux-command-line/
    * https://www.theknowledgeacademy.com/blog/unix-commands-cheat-sheet/
    
These aren't necessarily the best resources. Feel free to search for better ones. Also, don't forget that Unix has built-in manual pages for all of its commands. Just type `man <command>` at the command prompt. Use the space-bar to scroll through the documentation and "q" to exit.


### Setup and Submission

Our course repository is public. The instructions here aim to have you setup a fork of the course repository. Unfortunately because you are forking a public repo, your fork will have to be public also. 

You should be familiar with git from the first semester of this course. I assume that you all have github accounts and have setup things to be able to [push to github using ssh](https://docs.github.com/en/github/authenticating-to-github/connecting-to-github-with-ssh). The instuctions here lead you to:

We'll overview what you will do before going through step by step instructions.

1. Setup:
    1. Fork the class repository. Some directions in [fork-a-repo](https://docs.github.com/en/github/getting-started-with-github/fork-a-repo).
    1. Create a directory on your personal system where you will keep all course materials.
    1. In that directory, clone your fork of the repository.
    1. Using `git remote`, set the upstream to be the class repo, so you can pull from the class and push to your fork.

1. Submission:
    1. Copy your solutions into the appropriate directory (e.g. into `Labs/Lab.2/`) and with appropriate filename `Lab.2.solution.ipynb'.
    1. Commit / push your solutions.
    1. Grant access to course instructors.

Below are step by step instructions with examples (including example directory naming convention). Feel free to modify things as you see fit. 

#### Setup
You should only need to follow this instructions once. Here are some useful git commands:

* Git help: `git help`
* Git remote help: `git help remote`
* Check remote status: `git remote -v`
* Add a remote: `git remote add <stream name> <repo URL>`
* Add a remove: `git remote remove <stream name>`

Steps:
1. In a browser, log into GitHub and navigate to the [course repository](https://github.com/UTA-DataScience/DATA3402.Fall.2024).
1. On the top right of the page, press the fork button to create a new fork into your own GitHub account.
1. After successful fork, you should find the browser showing your fork of the course repository. Use the green "Code" button to copy path to the repo into your the clipboard of your computer.
1. Open a shell on your personal computer.
1. If you have not done so already, create a new directory/folder where you will keep all course material to navigate to it. For example: `mkdir Data-3402` and `cd Data-3402`.
1. Clone your fork of the repository using `git clone` followed by the path you copied into your clipboard. (copy/paste)
1. Paste the URL to your fork in the worksheet for the TAs and instructors.
1. Now go into the directory of your clone (`cd DATA3402.Fall.2024`).
1. Type `git remote -v` to see the current setup for fetch and pull.
1. Note the URL you see. This should be the same as what you used for your clone for both push and fetch.
1. Delete the origin remote using `git remote remove origin`.
1. Add the course repo as your remote using `git remote add origin https://github.com/UTA-DataScience/DATA3402.Fall.2024.git`.
1. Change the push to point to your fork. This means you will need the URL to your clone we copied earlier and confirmed as the original origin. The command will look something like: `git remote set-url --push origin  https://github.com/XXXXXX/DATA3402.Fall.2024.git`, where XXXXX is your username on GitHub.
1. Note that if you setup everything correctly, you now should be able to do `git pull` to get updates from the course repo, and do `git push` to push your commits into your own fork.

### Submission
These instructions outline how you submit files. Some useful commands:
* To add a file to local repository: `git add <file>`.
* To commit all changed files into local repository: `git -a -m "A message"`. You need to provide some comment when you commit. 
* To push the commited files from the local repository to GitHub: `git push`.
* To get updates from GitHub: `git pull`.

Steps:
1. To submit your labs, navigate to your clone of your fork of the course repository. 
1. Use `git pull` to make sure you have the latest updates. 
1. Make sure your copy of the lab your are working on is in the appropriate place in this clone. That means if you have the file elsewhere, copy it to the same directory in your clone of your fork. 
1. Note that in order to avoid future conflicts, you should always name your solution differently than the original file in the class repo. For example if your file is still named `Lab.2.ipynb` you should rename it using the `mv` command: `mv Lab.2.ipynb Lab.2.solution.ipynb`. 
1. Add and files you wish to submit into the repo. For example: `git add Labs/Lab.2/Lab.2.solution.ipynb`
1. Commit any changes: `git commit -a -m "Lab 2 updates"`
1. Push your changes: `git push`
1. Check on github website that your solutions have been properly submitted.

Before you leave the session today, submit your Lab.2 solution as is now to make sure you can do it. If you need to work further on your lab, navigate jupyter to the copy of the lab you just submitted and work there. Once done, repeat the commit and push commands to submit your updated solution. Note that lab 2 is due by midnight Monday  9/8/2024.


### Uniform Distribution
Lets start with generating some fake random data. You can get a random number between 0 and 1 using the python random module as follow:

In [1]:
import random
x=random.random()
print("The Value of x is", x)

The Value of x is 0.6127475031312778


Everytime you call random, you will get a new number.

*Exercise 1:* Using random, write a function `generate_uniform(N, mymin, mymax)`, that returns a python list containing N random numbers between specified minimum and maximum value. Note that you may want to quickly work out on paper how to turn numbers between 0 and 1 to between other values. 

In [2]:
# Skeleton
def generate_uniform(N,x_min,x_max):
    out = []
    ### BEGIN SOLUTION
    for _ in range(N):
        x =random.random()
        x_scaled = x * (x_max - x_min) + x_min
        out.append(x_scaled)
    
    ### END SOLUTION
    return out

In [3]:
# Test your solution here
data=generate_uniform(1000,-10,10)
print ("Data Type:", type(data))
print ("Data Length:", len(data))
if len(data)>0: 
    print ("Type of Data Contents:", type(data[0]))
    print ("Data Minimum:", min(data))
    print ("Data Maximum:", max(data))

Data Type: <class 'list'>
Data Length: 1000
Type of Data Contents: <class 'float'>
Data Minimum: -9.965709643263486
Data Maximum: 9.984953432348625


*Exercise 2a:* 
Write a function that computes the mean of values in a list. Recall the equation for the mean of a random variable $\bf{x}$ computed on a data set of $n$ values $\{ x_i \} = \{x_1, x_2, ..., x_n\}$  is ${\bf\bar{x}} = \frac{1}{n} \sum_i^n x_i$.

In [4]:
# Skeleton
def mean(Data):
    m=0 #total sum
    
    ### BEGIN SOLUTION
     
    if len(Data) == 0:
        return 0
    
    total = sum(Data)
        
    l = len(Data) #find length
    
    mean_value = total/l #find mean by diving total sum by length of list    
    
    ### END SOLUTION
    
    return mean_value

In [5]:
# Test your solution here
print ("Mean of Data:", mean(data))

Mean of Data: 0.06945018121746706


*Exercise 2b:* 
Write a function that computes the variance of values in a list. Recall the equation for the variance of a random variable $\bf{x}$ computed on a data set of $n$ values $\{ x_i \} = \{x_1, x_2, ..., x_n\}$  is ${\bf\langle x \rangle} = \frac{1}{n} \sum_i^n (x_i - {\bf\bar{x}})$.

In [6]:
# Skeleton
def variance(Data):
    m=0
    
    ### BEGIN SOLUTION
    
    #use what we had previously to find mean
    if len(Data) == 0:
        return 0
    
    total = sum(Data)
        
    l = len(Data) #find length
    
    mean_value = total/l    
    
    #sum of squared differences
    diff_squared = sum((x - mean_value)** 2 for x in Data)
    m = diff_squared/len(Data)
    
    ### END SOLUTION
    return m

In [7]:
# Test your solution here
print ("Variance of Data:", variance(data))

Variance of Data: 32.961463691526035


## Histogramming

*Exercise 3:* Write a function that bins the data so that you can create a histogram. An example of how to implement histogramming is the following logic:

* User inputs a list of values `x` and optionally `n_bins` which defaults to 10.
* If not supplied, find the minimum and maximum (`x_min`,`x_max`) of the values in x.
* Determine the bin size (`bin_size`) by dividing the range of the function by the number of bins.
* Create an empty list of zeros of size `n_bins`, call it `hist`.
* Loop over the values in `x`
    * Loop over the values in `hist` with index `i`:
        * If x is between `x_min+i*bin_size` and `x_min+(i+1)*bin_size`, increment `hist[i].` 
        * For efficiency, try to use continue to goto the next bin and data point.
* Return `hist` and the list corresponding of the bin edges (i.e. of `x_min+i*bin_size`).    

In [8]:
# Solution
def histogram(x,n_bins=10,x_min=None,x_max=None):
    ### BEGIN SOLUTION
    if x_min is None:
        x_min = min(x)
    if x_max is None:
        x_max = max(x)
        
    #bin_size
    bin_size = (x_max-x_min)/n_bins
    #empty list of zeros of size n_bins
    hist = [0] * n_bins
    bin_edges = [x_min + i * bin_size for i in range(n_bins + 1)]
    
    #Loop over the values in x
    for value in x:
        for i in range(n_bins):
            if x_min + i * bin_size <= value < x_min + (i+1) * bin_size:
                hist[i] += 1
                break 
        
    ### END SOLUTION

    return hist,bin_edges

In [9]:
# Test your solution here
h,b=histogram(data,100)
print(h)

[8, 12, 7, 12, 8, 7, 8, 14, 8, 9, 9, 9, 7, 11, 9, 6, 11, 12, 13, 6, 11, 4, 13, 12, 14, 6, 14, 19, 11, 12, 8, 7, 13, 8, 14, 13, 13, 14, 10, 14, 9, 5, 9, 4, 14, 10, 11, 10, 10, 8, 7, 6, 15, 10, 11, 3, 9, 9, 12, 11, 10, 9, 8, 6, 11, 5, 16, 13, 11, 9, 5, 11, 6, 9, 13, 10, 10, 13, 21, 11, 12, 7, 7, 7, 9, 10, 12, 9, 10, 9, 9, 14, 4, 6, 5, 11, 15, 13, 11, 14]


*Exercise 4:* Write a function that uses the histogram function in the previous exercise to create a text-based "graph". For example the output could look like the following:
```
[  0,  1] : ######
[  1,  2] : #####
[  2,  3] : ######
[  3,  4] : ####
[  4,  5] : ####
[  5,  6] : ######
[  6,  7] : #####
[  7,  8] : ######
[  8,  9] : ####
[  9, 10] : #####
```

Where each line corresponds to a bin and the number of `#`'s are proportional to the value of the data in the bin. 

In [10]:
# Solution
def draw_histogram(x,n_bins,x_min=None,x_max=None,character="#",max_character_per_line=20):
    ### BEGIN SOLUTION
    hist, bin_edges = histogram(x,n_bins,x_min, x_max)
    count_max = max(hist)
    
    #create the written graph
    for i in range(n_bins):
        bin_begin = bin_edges[i]
        bin_end =  bin_edges[i+1]
        count_reg = hist[i]
        # scale th number of character symbols based on the bin count
        num_character = int((count_reg / count_max) * max_character_per_line)
        print(f"[{bin_begin:.2f}, {bin_end:.2f}] : {character * num_character}")
    
    ### END SOLUTION

    return hist,bin_edges

In [11]:
# Test your solution here
h,b=draw_histogram(data,20)

[-9.97, -8.97] : ##############
[-8.97, -7.97] : ##############
[-7.97, -6.97] : #############
[-6.97, -5.98] : ##############
[-5.98, -4.98] : ################
[-4.98, -3.98] : ###################
[-3.98, -2.98] : ###############
[-2.98, -1.99] : ###################
[-1.99, -0.99] : ############
[-0.99, 0.01] : ###############
[0.01, 1.01] : ###############
[1.01, 2.00] : #############
[2.00, 3.00] : #############
[3.00, 4.00] : ################
[4.00, 5.00] : #############
[5.00, 5.99] : ####################
[5.99, 6.99] : ############
[6.99, 7.99] : ###############
[7.99, 8.99] : ###########
[8.99, 9.98] : ###################


## This doesn't acurately show the histogram becasue the bases of the symbols/ the colons don't line up :/

## Functional Programming

*Exercise 5:* Write a function the applies a booling function (that returns true/false) to every element in data, and return a list of indices of elements where the result was true. Use this function to find the indices of entries greater than 0.5. 

In [12]:
def where(mylist,myfunc):
    out= []
    
    ### BEGIN SOLUTION

    for index, value in enumerate(mylist):
        if myfunc(value):
            out.append(index)
    
    ### END SOLUTION
    
    return out

In [13]:
# Test your solution here
indices = where(data, lambda x: x > 0.5)  #find indices of entireis greater than 0.5
print("Indices of entries greater than 0.5:", indices)

Indices of entries greater than 0.5: [0, 2, 3, 11, 12, 15, 17, 18, 23, 25, 27, 28, 29, 30, 31, 32, 36, 37, 41, 43, 47, 48, 49, 50, 54, 56, 61, 63, 67, 68, 69, 72, 76, 77, 78, 82, 84, 87, 90, 91, 92, 94, 95, 96, 97, 100, 103, 107, 110, 112, 113, 118, 119, 120, 121, 122, 123, 124, 126, 130, 132, 133, 136, 141, 145, 146, 147, 148, 154, 156, 162, 163, 164, 169, 171, 172, 173, 175, 177, 180, 183, 184, 185, 189, 191, 192, 196, 197, 198, 199, 202, 203, 207, 211, 217, 219, 220, 221, 224, 225, 226, 229, 234, 235, 236, 237, 238, 240, 244, 245, 248, 249, 250, 251, 252, 254, 256, 261, 264, 267, 269, 270, 271, 272, 273, 275, 276, 279, 281, 282, 283, 286, 288, 291, 293, 296, 300, 301, 302, 303, 304, 306, 308, 316, 317, 320, 323, 324, 325, 327, 328, 332, 337, 339, 345, 346, 349, 350, 356, 357, 360, 363, 364, 366, 368, 370, 374, 375, 377, 379, 382, 383, 385, 388, 389, 392, 393, 395, 398, 399, 401, 402, 403, 406, 407, 409, 412, 415, 417, 419, 420, 424, 426, 429, 432, 433, 434, 435, 438, 439, 440, 445, 

### Wow, thats of lot of indices!

*Exercise 6:* The `inrange(mymin,mymax)` function below returns a function that tests if it's input is between the specified values. Write corresponding functions that test:
* Even
* Odd
* Greater than
* Less than
* Equal
* Divisible by

In [14]:
def in_range(mymin,mymax):
    def testrange(x):
        return x<mymax and x>=mymin
    return testrange

# Examples:
F1=in_range(0,10)
F2=in_range(10,20)

# Test of in_range
print (F1(0), F1(1), F1(10), F1(15), F1(20))
print (F2(0), F2(1), F2(10), F2(15), F2(20))

print ("Number of Entries passing F1:", len(where(data,F1)))
print ("Number of Entries passing F2:", len(where(data,F2)))

True True False False False
False False True True False
Number of Entries passing F1: 494
Number of Entries passing F2: 0


In [15]:
### BEGIN SOLUTION
def is_even():
    def even_test(x):
        return x % 2 == 0
    return even_test

def is_odd():
    def odd_test(x):
        return x % 2 != 0
    return odd_test

def greater_than(myval):
    def test_greater(x):
        return x > myval
    return test_greater

def less_than(myval):
    def test_less(x):
        return x < myval
    return test_less

def equal_to(myval):
    def test_equal(x):
        return x == myval
    return test_equal

def divisible_by(myval):
    def test_divisible(x):
        return x % myval == 0
    return test_divisible
    
    
### END SOLUTION

In [16]:
# Test your solution
F3 = is_even()
F4 = is_odd()
F5 = greater_than(7)
F6 = less_than(5)
F7 = equal_to(8)
F8 = divisible_by(3)

print(F3(8), F3(5))  
print(F4(6), F4(7))  
print(F5(6), F5(8))  
print(F6(4), F6(5), F6(6))  
print(F7(7), F7(8), F7(9))  
print(F8(10), F8(12), F8(15))  


print ("Number of Entries Even:", len(where(data,F3)))
print ("Number of Entries Odd:", len(where(data,F4)))
print ("Number of Entries Greater than 7:", len(where(data,F5)))
print ("Number of Entries Less than 5:", len(where(data,F6)))
print ("Number of Entries Equal to 8:", len(where(data,F7)))
print ("Number of Entries Divisible by 3:", len(where(data,F8)))



True False
False True
False True
True False False
False True False
False True True
Number of Entries Even: 0
Number of Entries Odd: 1000
Number of Entries Greater than 7: 152
Number of Entries Less than 5: 741
Number of Entries Equal to 8: 0
Number of Entries Divisible by 3: 0


*Exercise 7:* Repeat the previous exercise using `lambda` and the built-in python functions sum and map instead of your solution above. 

In [17]:
### BEGIN SOLUTION

def is_even():
    return lambda x: x % 2 == 0

def is_odd():
    return lambda x: x % 2 != 0

def greater_than(myval):
    return lambda x: x > myval

def less_than(myval):
    return lambda x: x < myval

def equal_to(myval):
    return lambda x: x == myval

def divisible_by(myval):
    return lambda x: x % myval == 0

def in_range(mymin, mymax):
    return lambda x: mymin <= x < mymax
    
### END SOLUTION

F1=in_range(0,15)
F2=in_range(15,30)

# Test of in_range
print (F1(0), F1(1), F1(10), F1(15), F1(20))
print (F2(0), F2(1), F2(10), F2(15), F2(20))

F3 = is_even()
F4 = is_odd()
F5 = greater_than(6)
F6 = less_than(4)
F7 = equal_to(9)
F8 = divisible_by(5)

print(F3(8), F3(5))  
print(F4(6), F4(7))  
print(F5(6), F5(8))  
print(F6(4), F6(5), F6(6))  
print(F7(7), F7(8), F7(9))  
print(F8(10), F8(12), F8(15))  

print("Number of Entries in range 0-15:", sum(map(F1, data)))
print("Number of Entries in range 15-30:", sum(map(F2, data)))
print("Number of Entries Even:", sum(map(F3, data)))
print("Number of Entries Odd:", sum(map(F4, data)))
print("Number of Entries Greater than 6:", sum(map(F5, data)))
print("Number of Entries Less than 4:", sum(map(F6, data)))
print("Number of Entries Equal to 9:", sum(map(F7, data)))
print("Number of Entries Divisible by 5:", sum(map(F8, data)))


True True True False False
False False False True True
True False
False True
False True
False False False
False False True
True False True
Number of Entries in range 0-15: 494
Number of Entries in range 15-30: 0
Number of Entries Even: 0
Number of Entries Odd: 1000
Number of Entries Greater than 6: 194
Number of Entries Less than 4: 697
Number of Entries Equal to 9: 0
Number of Entries Divisible by 5: 0


#### I don't know if that worked very well; if it didn't please let me know.

## Monte Carlo

*Exercise 7:* Write a "generator" function called `generate_function(func,x_min,x_max,N)`, that instead of generating a flat distribution, generates a distribution with functional form coded in `func`. Note that `func` will always be > 0.  

Use the test function below and your histogramming functions above to demonstrate that your generator is working properly.

Hint: A simple, but slow, solution is to a draw random number `test_x` within the specified range and another number `p` between the `min` and `max` of the function (which you will have to determine). If `p<=function(test_x)`, then place `test_x` on the output. If not, repeat the process, drawing two new numbers. Repeat until you have the specified number of generated numbers, `N`. For this problem, it's OK to determine the `min` and `max` by numerically sampling the function.  

In [18]:
def generate_function(func,x_min,x_max,N=1000):
    out = list()
    ### BEGIN SOLUTION
    x_samples = [x_min + (x_max - x_min) * i / 999 for i in range(1000)]
    y_samples = [func(x) for x in x_samples]
    y_min = min(y_samples)
    y_max = max(y_samples)
    
    while len(out) < N:
        test_x = random.uniform(x_min, x_max)
        p = random.uniform(y_min, y_max)

        #if p is less than or equal to the func value at test_c
        if p <= func(test_x):
            out.append(test_x)
    
    ### END SOLUTION
    
    return out

In [22]:
# A test function
def test_func(x,a=1,b=1):
    return abs(a*x+b)

def equal_to(myval, epsilon=1e-2):
    return lambda x: abs(x - myval) < epsilon

# Parameters for the test
x_min = -10
x_max = 10
N = 1000

#generate samples
samps = generate_function(test_func, x_min, x_max, N)

#filtering functions
odd_count = sum(map(is_odd(), samps))
greater_than_count = sum(map(greater_than(5), samps))
equal_to_count = sum(map(equal_to(7), samps))

# Print the results
print("Number of odd samples:", odd_count)
print("Number of samples greater than 5:", greater_than_count)
print("Number of samples equal to 7:", equal_to_count)

Number of odd samples: 1000
Number of samples greater than 5: 398
Number of samples equal to 7: 1


*Exercise 8:* Use your function to generate 1000 numbers that are normal distributed, using the `gaussian` function below. Confirm the mean and variance of the data is close to the mean and variance you specify when building the Gaussian. Histogram the data. 

In [25]:
import math

def gaussian(mean, sigma):
    def f(x):
        return math.exp(-((x-mean)**2)/(2*sigma**2))/math.sqrt(math.pi*sigma)
    return f

# Example Instantiation
g1=gaussian(0,1)
g2=gaussian(10,3)

In [26]:
# Generate 1000 samples
samples = generate_function(g1, -10, 10, 1000)

mean_sample = sum(samples) / len(samples)
variance_sample = sum((x - mean_sample) ** 2 for x in samples) / len(samples)

print("Generated Sample Mean:", mean_sample)
print("Generated Sample Variance:", variance_sample)

Generated Sample Mean: 0.027274046949496828
Generated Sample Variance: 0.9420606896540381


*Exercise 9:* Combine your `generate_function`, `where`, and `in_range` functions above to create an integrate function. Use your integrate function to show that approximately 68% of Normal distribution is within one variance.

In [30]:
def integrate(func, x_min, x_max, n_points=1000):
    samples = generate_function(func, x_min, x_max, n_points)
    # Define the range within one stdev
    condition = in_range(mean - sigma, mean + sigma)
    
    count_within_range = len(where(samples, condition))
    # Approximate the integral (area under the curve)
    integral = count_within_range / len(samples)
    
    return integral  #integral = area under the curve
    
    return integral


In [32]:
#test
mean = 0
sigma = 1
g1 = gaussian(mean, sigma)
# Perform integration 
area = integrate(g1, mean - 4 * sigma, mean + 4 * sigma)  # Wider range to ensure coverage

print("Approximate area under the curve within one degree pf standard deviation:", area)

Approximate area under the curve within one degree pf standard deviation: 0.679


## Here we see the approximate area under the curve within one degree of standard deviation is around 67.9%